### PR TITLE
[NPU] - Port plugin changes for dynamic shapes support

### PR DESCRIPTION
**Pull request by DariaMityagina**
### Details:
- Added `shape` metadata
- Shape tensors have a special prefix to distinguish shape tensors from others.

### Tickets:
 - E-115058

### CODE REVIEW

File: src/plugins/intel_npu/src/backend/src/zero_infer_request.cpp [hunk 1]
```diff
@@ -186,8 +221,20 @@ void ZeroInferRequest::get_result() {
...
 
     for (const auto& name : _outputAndStateOutputNames) {
-        const std::shared_ptr<ov::ITensor>& outputTensor = _allTensors.at(name);
-        const std::shared_ptr<ov::ITensor>& wrapperOutputTensor = _copyAllTensors.at(name);
+        const auto& outputTensor = _allTensors.at(name);
+        const auto& wrapperOutputTensor = _copyAllTensors.at(name);
+
+        if (isShapeTensorName(name)) {
+            const auto actualTensorName = name.substr(SHAPE_TENSOR_PREFIX.size());
+            ov::Shape actualDims;
```
[the remark this answers is not in the document]
Comment (DariaMityagina) [reply src/plugins/intel_npu/src/backend/src/zero_infer_request.cpp:229]: Done, thanks! https://github.com/openvinotoolkit/openvino/pull/23814/commits/50a89f534d371f1950a476fcbd03d3e6828bc1fc

File: src/plugins/intel_npu/src/al/include/intel_npu/al/icompiler.hpp [hunk 2]
```diff
@@ -51,10 +51,12 @@ struct NetworkMetadata final {
     std::vector<std::string> inputNames;
     std::vector<std::string> outputNames;
     std::vector<std::string> stateNames;
+    std::vector<std::string> shapeNames;
```
[the remark this answers is not in the document]
Comment (DariaMityagina) [reply src/plugins/intel_npu/src/al/include/intel_npu/al/icompiler.hpp:54]: No, we do not need CiD changes at this point.

[the remark this answers is not in the document]
Comment (ArtemySkrebkov) [reply src/plugins/intel_npu/src/al/include/intel_npu/al/icompiler.hpp:54]: Ok, I see, we can use Compiler n Plugin for testing indeed.

**Review by ArtemySkrebkov — APPROVED** [review]

**Comment by DariaMityagina** [comment]
OV integration job: https://vpux-ci-vpuip.ir.intel.com/job/ov_integration/job/ov_integration/2045/

File: src/plugins/intel_npu/src/al/src/sync_infer_request.cpp [hunk 3]
```diff
@@ -155,22 +168,23 @@ void SyncInferRequest::check_tensors() const {
...
-        tensor = ov::make_tensor(descriptor.precision, descriptor.transposedShape.get_shape());
+        tensor = ov::make_tensor(descriptor.precision, descriptor.transposedShape.get_max_shape());
     }
 
-    if (!isState) {
-        _copyAllTensors[tensorName] = std::move(tensor);
-        _allTensors[tensorName] = _copyAllTensors[tensorName];
-    } else {
+    if (tensorType == TensorType::Shape) {
+        _shapesTensors[tensorName] = tensor;
```
[the remark this answers is not in the document]
Comment (ArtemySkrebkov) [reply src/plugins/intel_npu/src/al/src/sync_infer_request.cpp:184]: Let's fix it in a separate PR, just not re-run CI.

**Pull request merged**